gef.py

```diff
@@ -5335,66 +5337,72 @@ def __iter__(self) -> Generator[str, None, None]:
 
         def __getitem__(self, structure_name: str) -> "ExternalStructureManager.Structure":
-            for structure in self.structures:
-                print(self.path, structure.name)
-                if structure.name == structure_name:
-                    return structure
-            raise KeyError(f"Structure {structure_name} not found")
+            return self.structures[structure_name]
 
     class Modules:
         def __init__(self, path: pathlib.Path) -> None:
-            self.modules: List[ExternalStructureManager.Module] = []
+            self.modules: Dict[str, ExternalStructureManager.Module] = {}
             self.root: pathlib.Path = path
 
             for entry in self.root.iterdir():
                 if not entry.is_file(): continue
                 if entry.suffix != ".py": continue
                 if entry.name == "__init__.py": continue
-                self.modules.append(ExternalStructureManager.Module(entry))
+                module = ExternalStructureManager.Module(entry)
+                self.modules[module.name] = module
             return
 
         def __contains__(self, structure_name: str) -> bool:
             """Return True if the structure name is found in any of the modules"""
             for module in self.modules:
-                if structure_name in [s.name for s in module.structures]:
+                if structure_name in module:
                     return True
             return False
 
         def __getitem__(self, module_name: str) -> "ExternalStructureManager.Module":
-            for module in self.modules:
-                if module_name == module.path.stem:
-                    return module
-            raise KeyError(f"Module {module_name} not found")
+            return self.modules[module_name]
 
         def __iter__(self):
             return iter(self.modules)
 
+    def __init__(self):
+        self.clear_caches()
+        return
+
+    def clear_caches(self):
+        self._path = None
+        self._modules = None
+        return
+
     @property
     def modules(self) -> "ExternalStructureManager.Modules":
-        return ExternalStructureManager.Modules(self.path)
+        if not self._modules:
+            self._modules = ExternalStructureManager.Modules(self.path)
+        return self._modules
 
     @property
     def path(self) -> pathlib.Path:
-        return pathlib.Path(gef.config["pcustom.struct_path"]).expanduser().absolute()
+        if not self._path:
+            self._path = pathlib.Path(gef.config["pcustom.struct_path"]).expanduser().absolute()
+        return self._path
 
     @property
     def structures(self) -> Generator[Tuple["ExternalStructureManager.Module", "ExternalStructureManager.Structure"], None, None]:
-        for module in self.modules:
-            for structure in module.structures:
-                yield module, structure
+        for module_name in self.modules:
+            module = self.modules[module_name]
```

Comment:
you can do `for module in self.modules.values():` or `for module_name, module in self.modules.items():`

Reply:
`self.modules` is not a `dict`

Reply:
Line 5347: `self.modules: Dict[str, ExternalStructureManager.Module] = {}`